rcl/include/rcl/guard_condition.h

```diff
@@ -0,0 +1,161 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__GUARD_CONDITION_H_
+#define RCL__GUARD_CONDITION_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcl/node.h"
+#include "rcl/visibility_control.h"
+
+/// Internal rcl guard condition implementation struct.
+struct rcl_guard_condition_impl_t;
+
+/// Handle for a rcl guard condition.
+typedef struct rcl_guard_condition_t
+{
+  struct rcl_guard_condition_impl_t * impl;
+} rcl_guard_condition_t;
+
+/// Options available for a rcl guard condition.
+typedef struct rcl_guard_condition_options_t
+{
+  /// Custom allocator for the guard condition, used for internal allocations.
+  rcl_allocator_t allocator;
+} rcl_guard_condition_options_t;
+
+/// Return a rcl_guard_condition_t struct with members set to NULL.
+RCL_PUBLIC
+rcl_guard_condition_t
+rcl_get_zero_initialized_guard_condition();
+
+/// Initialize a rcl guard_condition.
+/* After calling this function on a rcl_guard_condition_t, it can be passed to
+ * rcl_wait() and then concurrently it can be triggered to wake-up rcl_wait().
+ *
+ * The given rcl_node_t must be valid and the resulting rcl_guard_condition_t
+ * is only valid as long as the given rcl_node_t remains valid.
+ *
+ * Expected usage:
+ *
+ *    #include <rcl/rcl.h>
+ *
+ *    rcl_node_t node = rcl_get_zero_initialized_node();
+ *    rcl_node_options_t node_ops = rcl_node_get_default_options();
+ *    rcl_ret_t ret = rcl_node_init(&node, "node_name", &node_ops);
+ *    // ... error handling
+ *    rcl_guard_condition_t guard_condition = rcl_get_zero_initialized_guard_condition();
+ *    ret = rcl_guard_condition_init(
+ *      &guard_condition, &node, rcl_guard_condition_get_default_options());
+ *    // ... error handling, and on shutdown do deinitialization:
+ *    ret = rcl_guard_condition_fini(&guard_condition, &node);
+ *    // ... error handling for rcl_guard_condition_fini()
+ *    ret = rcl_node_fini(&node);
+ *    // ... error handling for rcl_deinitialize_node()
+ *
+ * This function does allocate heap memory.
+ * This function is not thread-safe.
+ * This function is lock-free.
+ *
+ * \TODO(wjwwood): does this function need a node to be passed to it? (same for fini)
```

Comment:
Currently `rmw_create_guard_condition` doesn't get any context passed. I think this should change in the future since we can't assume that every rmw impl. can perform (any) operation without context. Therefore I would argue to keep passing the node here (and sometime later do the same in the rmw interface).